src/app/AttributePathExpandIterator.h

```diff
@@ -20,78 +20,100 @@
-
-    /**
-     * Proceed the iterator to the next attribute path in the given cluster info.
-     *
-     * Returns false if AttributePathExpandIteratorDataModeDataModel has exhausted all paths in the given AttributePathParams list.
-     */
-    bool Next();
-
-    /**
-     * Fills the aPath with the path the iterator currently points to.
-     * Returns false if the iterator is not pointing to a valid path (i.e. it has exhausted the cluster info).
-     */
-    bool Get(ConcreteAttributePath & aPath)
+    class State
     {
-        aPath = mOutputPath;
-        return (mpAttributePath != nullptr);
-    }
+    public:
+        friend class AttributePathExpandIterator;
 
-    /**
-     * Reset the iterator to the beginning of current cluster if we are in the middle of expanding a wildcard attribute id for some
-     * cluster.
-     *
-     * When attributes are changed in the middle of expanding a wildcard attribute, we need to reset the iterator, to provide the
-     * client with a consistent state of the cluster.
-     */
-    void ResetCurrentCluster();
-
-    /** Start iterating over the given `paths` */
-    inline void ResetTo(SingleLinkedListNode<AttributePathParams> * paths)
-    {
-        *this = AttributePathExpandIterator(mDataModelProvider, paths);
-    }
+        /// External callers can only ever start iterating on a new path from the beginning
+        static State StartIterating(SingleLinkedListNode<AttributePathParams> * path) { return State(path); }
+
+        /// Copies are allowed
+        State(const State &)             = default;
+        State & operator=(const State &) = default;
+
+        State() : mAttributePath(nullptr) {}
+
+        /// Reset the iterator to the beginning of current cluster if we are in the middle of expanding a wildcard attribute id for
+        /// some cluster.
+        ///
+        /// When attributes are changed in the middle of expanding a wildcard attribute, we need to reset the iterator, to provide
+        /// the client with a consistent state of the cluster.
+        void IterateFromTheStartOfTheCurrentCluster()
+        {
+            VerifyOrReturn(mAttributePath != nullptr && mAttributePath->mValue.HasWildcardAttributeId());
+            mLastOutputPath.mAttributeId = kInvalidAttributeId;
+            mLastOutputPath.mExpanded    = true;
+        }
+
+    protected:
+        State(SingleLinkedListNode<AttributePathParams> * path) :
+            mAttributePath(path), mLastOutputPath(kInvalidEndpointId, kInvalidClusterId, kInvalidAttributeId)
+        {
+            mLastOutputPath.mExpanded = true;
+        }
+
+        SingleLinkedListNode<AttributePathParams> * mAttributePath;
+        ConcreteAttributePath mLastOutputPath;
+    };
+
+    AttributePathExpandIterator(DataModel::Provider * dataModel, State & state) : mDataModelProvider(dataModel), mState(state) {}
+
+    // this class may not be copied. A new one should be created when needed and they
```

Comment:
nit: Lots of comments don't have starting capital at start of sentence, and missing period at the end. Consider cleaning that up.

Reply:
I updated several comments now. I would say however that this comment requirement (uppercase and especially ending with a period is currently *not* a SDK style guide ... especially the "ends with period" is probably not followed in several files).